### DISCUSSION TITLE
We're having TypeScript server performance issues

### DISCUSSION THREAD

**Discussion by divmgl** [TanStack/router#1518] [Q&A]
Hello, since introducing TanStack Router into our app we're having some TypeScript performance issues where a couple of saves or changes to the routes will cause the TypeScript server to slow to a crawl. Has anyone run into this? Only a full restart of VSCode solves the problem.

**Comment by chorobin** [comment]
Can you run `tsc --extendedDiagnostics` to see how long it takes checking. Are you using `zod`, even before router?

**Reply by divmgl** [reply]
We're using tRPC and using Zod for inputs. We are also seeing very big slowdowns with tRPC.

I will take a look at what `tsc` brings back.

**Reply by divmgl** [reply]
Here are the results. Not sure how to action on this.

```
Files:                         6537
Lines of Library:             40241
Lines of Definitions:        777922
Lines of TypeScript:         169121
Lines of JavaScript:              0
Lines of JSON:                  170
Lines of Other:                   0
Identifiers:                1105171
Symbols:                    1826131
Types:                       633185
Instantiations:             5496814
Memory used:               2240346K
Assignability cache size:    335906
Identity cache size:          37852
Subtype cache size:           37068
Strict subtype cache size:    33071
I/O Read time:                0.22s
Parse time:                   1.53s
ResolveModule time:           0.66s
ResolveTypeReference time:    0.02s
ResolveLibrary time:          0.00s
Program time:                 2.87s
Bind time:                    0.88s
Check time:                  18.56s
printTime time:               0.45s
Emit time:                    0.45s
I/O Write time:               0.01s
Total time:                  22.76s
```

**Reply by chorobin** [reply]
Yeah, that's quite slow. I've seen router with a combination of `zod` slow things down. It's on my to-do list to try to optimise. Do you use any other typescript validation libraries?

**Reply by divmgl** [reply]
Nope, that is it. We only use Zod. We tried using Typebox but the API was not a 1:1 match with Zod and we had to write a lot of custom code to glue things up. We've tried a variety of different things but we're in a monorepo configuration so our options are limited. We also are not using composite projects and instead have everything in a single TypeScript configuration at the root, for simplicity's sake. 

But our app is really not that big and the slowdowns with Zod are substantial. The issue however is that now with TanStack Router also in our project we're seeing some `tsc` timeouts which cause us to have to restart VSCode.

**Reply by chorobin** [reply]
I'm going to be doing some profiling with zod and router so I hope I will be able to improve things a bit

**Reply by schiller-manuel** [reply]
Just to make sure that the slowdown comes from zod, can you replace the usage of zod by some simple dummy methods and re-run `tsc --extendedDiagnostics `?

**Reply by divmgl** [reply]
We use Zod in hundreds of places so I don't think this will be possible without completely breaking the app, which would then result in additional slowdowns I believe.

**Reply by divmgl** [reply]
I think there's a way to capture a heatmap of what's taking the longest, I'll see if I can do this.
